plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoMetadata.java

```diff
@@ -603,11 +607,41 @@ public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(C
                 handle.getRemoteTableName(),
                 handle.getFilter(),
                 newDomain,
+                handle.getProjectedColumns(),
                 handle.getLimit());
 
         return Optional.of(new ConstraintApplicationResult<>(handle, remainingFilter, false));
     }
 
+    @Override
+    public Optional<ProjectionApplicationResult<ConnectorTableHandle>> applyProjection(
+            ConnectorSession session,
+            ConnectorTableHandle handle,
+            List<ConnectorExpression> projections,
+            Map<String, ColumnHandle> assignments)
+    {
+        MongoTableHandle mongoTableHandle = (MongoTableHandle) handle;
+
+        Set<ColumnHandle> newColumns = ImmutableSet.copyOf(assignments.values());
+
+        if (mongoTableHandle.getProjectedColumns().isPresent() && newColumns.equals(mongoTableHandle.getProjectedColumns().get())) {
+            return Optional.empty();
+        }
+
+        // TODO: support dereference pushdown
+        ImmutableSet.Builder<MongoColumnHandle> projectedColumns = ImmutableSet.builder();
+        ImmutableList.Builder<Assignment> assignmentList = ImmutableList.builder();
+        assignments.forEach((name, column) -> {
+            MongoColumnHandle columnHandle = (MongoColumnHandle) column;
+            projectedColumns.add(columnHandle);
+            assignmentList.add(new Assignment(name, column, columnHandle.getType()));
+        });
```

Comment:
I think you can prepare `projectedColumns` first, and then `assignmentList` from `projectedColumns`. This will look more linear.